src/client/dfuse/il/intercept.h

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#ifndef __INTERCEPT_H__
+#define __INTERCEPT_H__
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/uio.h>
+#include "dfuse_log.h"
+#include "dfuse_gah.h"
+#include "dfuse_fs.h"
+#include "ioil_io.h"
+#include "ioil_api.h"
+
+/* Low level I/O functions we intercept
+ *
+ * We purposefully skip the following:
+ * fileno
+ * fileno_unlocked
+ * sync
+ * munmap
+ * msync
+ * mremap
+ * select
+ * all aio routines (for now)
+ * fcntl (for now though we likely need for dup)
+ */
+#define FOREACH_ALIASED_INTERCEPT(ACTION)                                     \
+	ACTION(FILE *,  fopen,     (const char *, const char *))              \
+	ACTION(FILE *,  freopen,   (const char *, const char *, FILE *))      \
+	ACTION(int,     open,      (const char *, int, ...))                  \
+	ACTION(ssize_t, pread,     (int, void *, size_t, off_t))              \
+	ACTION(ssize_t, pwrite,    (int, const void *, size_t, off_t))        \
+	ACTION(off_t,   lseek,     (int, off_t, int))                         \
+	ACTION(ssize_t, preadv,    (int, const struct iovec *, int, off_t))   \
+	ACTION(ssize_t, pwritev,   (int, const struct iovec *, int, off_t))   \
+	ACTION(void *,  mmap,      (void *, size_t, int, int, int, off_t))
+
+#define FOREACH_SINGLE_INTERCEPT(ACTION)                                      \
+	ACTION(int,     fclose,    (FILE *))                                  \
+	ACTION(int,     close,     (int))                                     \
+	ACTION(ssize_t, read,      (int, void *, size_t))                     \
+	ACTION(ssize_t, write,     (int, const void *, size_t))               \
+	ACTION(ssize_t, readv,     (int, const struct iovec *, int))          \
+	ACTION(ssize_t, writev,    (int, const struct iovec *, int))          \
+	ACTION(int,     fsync,     (int))                                     \
+	ACTION(int,     fdatasync, (int))                                     \
+	ACTION(int,     dup,       (int))                                     \
+	ACTION(int,     dup2,      (int, int))                                \
+	ACTION(int,     fcntl,     (int fd, int cmd, ...))                    \
+	ACTION(FILE *,  fdopen,    (int, const char *))
+
+#define FOREACH_INTERCEPT(ACTION)            \
+	FOREACH_SINGLE_INTERCEPT(ACTION)     \
+	FOREACH_ALIASED_INTERCEPT(ACTION)
+
+#ifdef IOIL_PRELOAD
+#include <dlfcn.h>
+
+#define IOIL_FORWARD_DECL(type, name, params)  \
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon